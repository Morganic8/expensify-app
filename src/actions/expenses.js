import uuid from 'uuid';
import database from '../firebase/firebase';
//component calls action generator
//action generator returns object
//component dispatches object
//redux store changes



//ASYNC
//component calls action generator
//action generator returns function
//component dispatches function (?)
//function runs (has the ability to dispatch other actions and do whatever it wants)


//action generator
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense

});

export const startAddExpense = (expenseData = {}) => {
    //thunk allows us to return a function and access dispatch
    return (dispatch) => {
        //defaults
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        }   = expenseData;

        //object to be used with "PUSH"
        const expense = {description, note, amount, createdAt};

        //send to firebase****
        //for the "THEN" call - for success cases "PUSH" allows us to access the "ref" value
       return database.ref('expenses').push(expense).then( (ref)=> {
            //update redux store****
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
};

export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

//EDIT EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})
