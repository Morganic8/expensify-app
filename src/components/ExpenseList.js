import React from 'react';
import ExpenseItem from './ExpenseListItem';

import { connect } from 'react-redux';

import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.map( (expense)=> 
            <ExpenseItem key={expense.id} {...expense}/>)
        }
    </div>
);

//Connect Component to store
//First Brackets is what you want from the store
//Second  Brackets is the component you want to connect with

const mapStateToProps = (state) => {
    //must return object
    //anything you return here can be added as a 'prop' above
    return {
         expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);

