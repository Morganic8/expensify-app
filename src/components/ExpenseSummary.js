import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';


export const ExpenseSummary = (props) =>(
    
    <div>
        <p>
            Viewing {props.expenses.length} 
            {props.expenses.length == 1 ? " expense " : " expenses "}
            that total to {numeral(props.expensesTotal / 100).format('$0,0.00')}
        </p>
    </div>
)

const mapStateToProps = (state, filters) => {
    return {
        expenses : selectExpenses(state.expenses, state.filters),
        expensesTotal: selectExpensesTotal(state.expenses, state.filters)
    }
}
export default connect(mapStateToProps)(ExpenseSummary);