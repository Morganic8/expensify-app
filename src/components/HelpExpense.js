import React from 'react';

const HelpExpense = () => (
    <div>
        I need help!
    </div>
);


export default HelpExpense;