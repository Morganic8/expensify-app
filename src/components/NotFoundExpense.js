import React from 'react';
import {Link} from 'react-router-dom';


const NotFoundExpense = () => (
    <div>
        404 - <Link to="/"> Go Home</Link>
    </div>
);

export default NotFoundExpense;