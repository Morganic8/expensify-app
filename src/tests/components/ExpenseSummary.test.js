import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';


test('should render Expense summary with expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={235}/>)

    expect(wrapper).toMatchSnapshot();
});

test('should render Expense summary with multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={23} expensesTotal={23642363265}/>)

    expect(wrapper).toMatchSnapshot();
});
