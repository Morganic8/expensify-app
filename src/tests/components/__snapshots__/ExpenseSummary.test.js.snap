// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense summary with expenses 1`] = `
<div>
  <p>
    Viewing 
    1
     
     expense 
     totalling 
    $2.35
  </p>
</div>
`;

exports[`should render Expense summary with multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    23
     
     expenses 
     totalling 
    $236,423,632.65
  </p>
</div>
`;
